src/libawkward/array/ListArray.cpp

```diff
@@ -353,7 +354,34 @@ namespace awkward {
 
   template <typename T>
   const std::shared_ptr<Content> ListArrayOf<T>::flatten(int64_t axis) const {
-    throw std::runtime_error("FIXME: not implemented");
+    if(axis <= -1)
+      throw std::invalid_argument("axis must be a non-negative integer (can't count from the end)");
+    int64_t lenstarts = starts_.length();
+    if (stops_.length() < lenstarts) {
+      util::handle_error(failure("len(stops) < len(starts)", kSliceNone, kSliceNone), classname(), identities_.get());
+    }
+
+    // FIXME: calculate new ranges
+    IndexOf<T> nextstarts(lenstarts);
+    IndexOf<T> nextstops(lenstarts);
+
+    // FIXME: pre-calculate length and make it big enough
+    Index64 toarray(content_.get()->length()*lenstarts);
+
+    int64_t lenarray(0);
+    struct Error err = util::awkward_listarray_flatten_64<T>(
+      nextstarts.ptr().get(),
+      nextstops.ptr().get(),
+      starts_.ptr().get(),
+      stops_.ptr().get(),
+      lenstarts,
+      toarray.ptr().get(),
+      &lenarray);
+    util::handle_error(err, classname(), identities_.get());
+
+    // FIXME: shrink it here
+    Index64 indxarray(toarray.ptr(), 0, lenarray);
+    return content_.get()->carry(indxarray);
```

Comment:
@jpivarski - ```content_.get()->carry(indxarray)``` is indeed what I use. 
I will split ```util::awkward_listarray_flatten``` on two parts as you suggest. So far it both calculates the length and fills in an index array...
The newly calculated ranges are not used here. I was thinking they might be needed for the extra-dimensions, however, I'll drop them for now.

BTW, the more complicated test works fine. It's the indices for the sliced one give me a problem I do not understand yet...

Reply:
I saw in your first kernel that you didn't pass the `Index64`'s `offset()` down anywhere. The `Index64` holds a shared pointer to its raw bytes. A simple slice does not copy those bytes; it just changes the `offset` (initially `0`) and `length` to the range we're considering, so that the sliced `Index64` is a view into the original one (sharing reference counting). This `offset` has to be explicitly passed into the kernel and used to get the correct values.

In general, you have to pass

```c++
index.ptr().get()   // pointer to the BASE of the array
index.offset()      // where this index really starts, relative to the base
index.length()      // the length of the index
```

into a kernel to use some `index`. In some cases, we've already verified that some lengths are equal to or less than others (especially `starts` and `stops`), so we might not pass all lengths down.

The kernels are a low-level interface, more like programming C than C++.

Reply:
You do use `carry`... Oh! I was confused because I thought that you were calculating new `nextstarts` and `nextstops`... But you are? Or you're not using them? The kernel you want should take the old `starts` and `stops` in (which means the base pointers, offsets for both, but only the length of `starts` because excess values in `stops` are ignored) and it should write a `nextcarry` out (you named it `toarray`). Conventionally, I've been putting the arrays that get written _first_ in the argument list. For C functions, I need conventions like that to stay sane.

Ignoring the `starts_.offset()` and `stops_.offset()` is very likely your error. But you also need to precalculate the length with a second kernel. There's no way to guess how big it can be from the `starts` length and the `content_` length: the fact that it can have gaps _and_ overlaps allows it to be arbitrarily big.

Reply:
Great, thanks!

Reply:
I'll restructure and rename the arguments to follow the conventions :-)
What about the function names? ```awkward_listarray_flatten_length``` to calculate the length and ```awkward_listarray_flatten``` to fill in the ```tocarry```?

Reply:
Those are good names. I start all kernel names with `awkward_` because C function names are a shared namespace, then `listarray` for the class and `flatten` for the operation.

I have the following conventions for kernel arguments:

   * scalars (like `length`) or arrays (like `tocarry`) that we're writing to first;
   * arrays that we're reading from (like `fromstarts`, `fromstops`) next, and I put `const` in the function signature;
   * scalars that we read from (like `startsoffset` and `stopsoffset`) last;
   * `template <typename C, typename T>` for a content size `C` on `fromstarts` and carry size `T` on `tocarry`;
   * we only specialize three content sizes `C`: `int32_t`, `uint32_t`, and `int64_t`; function names are `awkward_listarray32_flatten_64`, `awkward_listarrayU32_flatten_64`, `awkward_listarray64_flatten_64`;
   * we only specialize one carry size: `int64_t`; the template is a placeholder; this is the `_64` at the end of each specialized function name;
   * unfortunately, it looks like the second tempating (on the libawkward side, in util.h/util.cpp) uses `template <typename T>` where `T` is the content size. It's also the content size when specializing the `ListArray` class (and others, like `ListOffsetArray`).

These conventions mitigate the downside of having a pure C interface between the two parts, but it wouldn't have been possible to write the Numba implementation without a pure C interface.